extensions/binding/org.eclipse.smarthome.binding.mqtt.generic/src/main/java/org/eclipse/smarthome/binding/mqtt/generic/internal/convention/homeassistant/HaID.java

```diff
@@ -37,14 +37,14 @@
      * Creates a {@link HaID} object for a given HomeAssistant MQTT topic.
      *
      * @param mqttTopic A topic like "homeassistant/binary_sensor/garden/config" or
-     *            "homeassistant/binary_sensor/0/garden/config"
+     *                      "homeassistant/binary_sensor/0/garden/config"
      */
     public HaID(String mqttTopic) {
         String[] strings = mqttTopic.split("/");
         if (strings.length < 3) {
             throw new IllegalArgumentException("MQTT topic not a HomeAssistant topic!");
         }
-        if (strings.length >= 4) {
+        if (strings.length >= 5) {
```

Comment:
So, we need equal to or more then five elements but we only access the second, third and fourth?

Reply:
Yes, that is right.
We have to be carefull because arrays are 0-based.
And we ignore the last element, because it always has to be 'config'

Reply:
Thanks for that info.

Would it make sense to check the last element or is this already known because of the topic and the call stack?

```Java
if (!"config".equals(strings[strings.length - 1])) {
    throw new IllegalArgumentException("Last part of MQTT topic must be equal to \"config\"!");
}
```

Reply:
All the places, that currently call this constructor have something like:
```
        if (!topic.endsWith("/config")) {
            return;
        }
```
before the call.

I wanted to do minimal changes for the problem I found.
Personally I would code it like this:

```
        if (strings.length < 4 || strings.length > 5) {
            throw new IllegalArgumentException("MQTT topic not a HomeAssistant topic (wrong length)!");
        }
        if (!"config".equals(strings[strings.length - 1])) {
            throw new IllegalArgumentException("Last part of MQTT topic must be equal to \"config\"!");
        }
        if (strings.length == 5) {
```
and also have tests for all those cases.
Would that also be your line of thinking?

Reply:
I agree that the changes should be minimal and fix the specific part.

The cleanup of the code could be another PR.
IMHO the code you posted is better to understand without reading all the call stack etc.